rootless-install.sh

```diff
@@ -0,0 +1,268 @@
+	fi
+
+
+	if [ ! -w "$XDG_RUNTIME_DIR" ]; then
+		if [ -n "$SYSTEMD" ]; then
+			>&2 echo "Aborting because systemd was detected but XDG_RUNTIME_DIR (\"$XDG_RUNTIME_DIR\") does not exist or is not writable"
+			>&2 echo "Hint: this could happen if you changed users with 'su' or 'sudo'. To work around this:"
+			>&2 echo "- try again by first running with root privileges 'loginctl enable-linger <user>' where <user> is the unprivileged user and export XDG_RUNTIME_DIR to the value of RuntimePath as shown by 'loginctl show-user <user>'"
+			>&2 echo "- or simply log back in as the desired unprivileged user (ssh works for remote machines)"
+			exit 1
+		fi
+		export XDG_RUNTIME_DIR="/tmp/docker-rootless-$(id -u)"
+		mkdir -p "$XDG_RUNTIME_DIR"
+	fi
+
+	# Already installed verification (unless force?). Only having docker cli binary previously shouldn't fail the build.
+	if [ -x "$BIN/$DAEMON" ]; then
+		# If rootless installation is detected print out the modified PATH and DOCKER_HOST that needs to be set.
+		echo "# Existing rootless Docker detected at $BIN/$DAEMON"
+		print_instructions
+	fi
+
+	INSTRUCTIONS=
+
+	if ! ( which newuidmap >/dev/null 2>&1 && test -u $(which newuidmap) ); then
+		if which apt-get >/dev/null 2>&1; then
+			INSTRUCTIONS='apt-get install -y uidmap\n'
+		elif which dnf >/dev/null 2>&1; then
+			INSTRUCTIONS='dnf install -y shadow-utils\n'
+		else
+			echo "Missing newuidmap or no setuid bit set. Please install with a package manager."
+			echo 1
+		fi
+	fi
+	
+	if [ -f /proc/sys/kernel/unprivileged_userns_clone ]; then
+		if [ "1" != "$(cat /proc/sys/kernel/unprivileged_userns_clone)" ]; then
+			INSTRUCTIONS="${INSTRUCTIONS}cat <<EOT > /etc/sysctl.d/50-rootless.conf
+kernel.unprivileged_userns_clone = 1
+EOT
+sysctl --system"
+		fi
+	fi
+
+	if [ -n "$INSTRUCTIONS" ]; then
+		echo "# Missing system requirements. Please run following commands to
+# install the requirements and run this installer again"
+
+		echo
+		echo "cat <<EOF | sudo sh -x"
+		echo "$INSTRUCTIONS"
+		echo "EOF"
+		echo
+		exit 1
+	fi
+
+	if ! grep "$(id -un)" /etc/subuid >/dev/null 2>&1; then
+		>&2 echo "Could not find records for the current user $(id -un) from /etc/subuid . Please make sure valid subuid range is set there."
+		exit 1
+	fi
+	if ! grep "$(id -gn)" /etc/subgid >/dev/null 2>&1; then
+		>&2 echo "Could not find records for the current user $(id -un) from /etc/subgid . Please make sure valid subuid range is set there."
+		exit 1
+	fi
+}
+
+start_docker() {
+	tmpdir=$(mktemp -d)
+	mkdir -p $tmpdir/lower $tmpdir/upper $tmpdir/work $tmpdir/merged
+	if "$BIN"/rootlesskit mount -t overlay overlay -olowerdir=$tmpdir/lower,upperdir=$tmpdir/upper,workdir=$tmpdir/work $tmpdir/merged >/dev/null 2>&1; then
+		USE_OVERLAY=1
+	fi
+	rm -rf "$tmpdir"
+
+
+	if [ -z "$SYSTEMD" ]; then
+		start_docker_nonsystemd
+		return
+	fi
+	
+	mkdir -p $HOME/.config/systemd/user
+	
+	DOCKERD_FLAGS="--experimental"
+	
+	if ! which iptables >/dev/null 2>&1 ; then
```

Comment:
@AkihiroSuda Even if iptables exists, the daemon fails to start because of permission errors when using iptables:
```
failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: iptables failed: iptables -t nat -N DOCKER: modprobe: ERROR: could not insert 'ip_tables': Operation not permitted
iptables v1.6.0: can't initialize iptables table `nat': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.
```

Separately from the above error, there are warnings that are due to the fact that libnetwork uses `modprobe` a lot:
```
WARN[2019-02-13T04:08:23.776052798Z] Running modprobe nf_nat failed with message: `modprobe: ERROR: could not insert 'nf_nat': Operation not permitted
insmod /lib/modules/4.4.0-141-generic/kernel/net/netfilter/nf_conntrack.ko`, error: exit status 1
WARN[2019-02-13T04:08:23.777322545Z] Running modprobe xt_conntrack failed with message: `modprobe: ERROR: could not insert 'xt_conntrack': Operation not permitted
insmod /lib/modules/4.4.0-141-generic/kernel/net/netfilter/x_tables.ko`, error: exit status 1
```
Not sure if that's important.

Reply:
I think user should ask admin to load these modules or set iptables=false manually

Reply:
@AkihiroSuda are you saying we should not even add `--iptables=false` if `iptables` does not exist ?

Reply:
@AkihiroSuda Could you explain the limitations to rootless when this flag is set? It is very common for this to fail but it isn't a blocker for starting the daemon. I would rather print a warning, allowing users to enable it later. The socat based port forwarding works with iptables false as well.

Reply:
I assumed hosts without iptables are insecure regardless to Docker and out of the first-class support, but perhaps `--iptables=false` is legit on nftables-based set up.